File: app/src/test/java/com/launchkey/android/authenticator/sdk/ui/internal/view/ExpirationTimerTest.java
/*
 *  Copyright (c) 2018. iovation, LLC. All rights reserved.
 */

package com.launchkey.android.authenticator.sdk.ui.internal.view;

import com.launchkey.android.authenticator.sdk.ui.internal.common.TimeAgo;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;

// Robolectric runner required to
// shadow a static call of Color.rgb() for
// default values in BaseExpirationTimer
@RunWith(RobolectricTestRunner.class)
@Config(sdk = 23)
public class ExpirationTimerTest {

    @Test
    public void testMillisToMinSecNullCheck() {

        final long[] minSec = ExpirationTimer.calculateMinSec(100);
        assertNotNull(minSec);
    }

    @Test
    public void testMillisToMinSecLengthCheck() {

        final long[] minSec = ExpirationTimer.calculateMinSec(100);
        assertEquals(2, minSec.length);
    }

    @Test
    public void testMillisToMinSecNegativeMillis() {

        final long[] minSec = ExpirationTimer.calculateMinSec(-100);
        assertNotNull(minSec);
        assertEquals(0L, minSec[0]);
        assertEquals(0L, minSec[1]);
    }

    @Test
    public void testMillisToMinSecBarelyBelowSixtySeconds() {

        final long[] minSec = ExpirationTimer.calculateMinSec(59999);
        assertNotNull(minSec);
        assertEquals(0L, minSec[0]);
        assertEquals(59L, minSec[1]);
    }

    @Test
    public void testMillisToMinSecSixtySeconds() {

        final long[] minSec = ExpirationTimer.calculateMinSec(60000);
        assertNotNull(minSec);
        assertEquals(1L, minSec[0]);
        assertEquals(0L, minSec[1]);
    }

    @Test
    public void testMillisToMinSecBarelyAboveSixtySeconds() {

        final long[] minSec = ExpirationTimer.calculateMinSec(60001);
        assertNotNull(minSec);
        assertEquals(1L, minSec[0]);
        assertEquals(0L, minSec[1]);
    }

    @Test
    public void testConvertMinSecToFormattedStringRegular() {

        final String timeFormatted = ExpirationTimer.convertMinSecToFormattedString(ExpirationTimer.FORMAT_DEFAULT, 1, 59);
        assertNotNull(timeFormatted);
        assertEquals("01:59", timeFormatted);
    }

    @Test
    public void testConvertMinSecToFormattedStringFiftyNineSeconds() {

        final String timeFormatted = ExpirationTimer.convertMinSecToFormattedString(ExpirationTimer.FORMAT_DEFAULT, 0, 59);
        assertNotNull(timeFormatted);
        assertEquals("00:59", timeFormatted);
    }

    @Test
    public void testConvertMinSecToFormattedStringDisallowSixtySeconds() {

        final String timeFormatted = ExpirationTimer.convertMinSecToFormattedString(ExpirationTimer.FORMAT_DEFAULT, 0, 60);
        assertNotNull(timeFormatted);
        assertEquals("00:00", timeFormatted);
    }

    @Test
    public void testConvertMinSecToFormattedStringDisallowOverSixtySeconds() {

        final String timeFormatted = ExpirationTimer.convertMinSecToFormattedString(ExpirationTimer.FORMAT_DEFAULT, 0, 61);
        assertNotNull(timeFormatted);
        assertEquals("00:00", timeFormatted);
    }

    @Test
    public void testConvertMinSecToFormattedStringDisallowUnderZeroSeconds() {

        final String timeFormatted = ExpirationTimer.convertMinSecToFormattedString(ExpirationTimer.FORMAT_DEFAULT, 0, -1);
        assertNotNull(timeFormatted);
        assertEquals("00:00", timeFormatted);
    }

    @Test
    public void testContentDescriptionOneSecond() {
        final String expected = "Expires in 1 second";
        final String actual = ExpirationTimer.getContentDescription(
                0,
                1,
                new TimeAgo(androidx.test.core.app.ApplicationProvider.getApplicationContext()),
                androidx.test.core.app.ApplicationProvider.getApplicationContext().getResources());
        assertEquals(expected, actual);
    }

    @Test
    public void testContentDescriptionFiveSeconds() {
        final String expected = "Expires in 5 seconds";
        final String actual = ExpirationTimer.getContentDescription(
                0,
                5,
                new TimeAgo(androidx.test.core.app.ApplicationProvider.getApplicationContext()),
                androidx.test.core.app.ApplicationProvider.getApplicationContext().getResources());
        assertEquals(expected, actual);
    }

    @Test
    public void testContentDescriptionOneMinute() {
        final String expected = "Expires in 1 minute";
        final String actual = ExpirationTimer.getContentDescription(
                1,
                0,
                new TimeAgo(androidx.test.core.app.ApplicationProvider.getApplicationContext()),
                androidx.test.core.app.ApplicationProvider.getApplicationContext().getResources());
        assertEquals(expected, actual);
    }

    @Test
    public void testContentDescriptionOneMinuteOneSecond() {
        final String expected = "Expires in 1 minute and 1 second";
        final String actual = ExpirationTimer.getContentDescription(
                1,
                1,
                new TimeAgo(androidx.test.core.app.ApplicationProvider.getApplicationContext()),
                androidx.test.core.app.ApplicationProvider.getApplicationContext().getResources());
        assertEquals(expected, actual);
    }

    @Test
    public void testContentDescriptionOneMinuteFortySeconds() {
        final String expected = "Expires in 1 minute and 40 seconds";
        final String actual = ExpirationTimer.getContentDescription(
                1,
                40,
                new TimeAgo(androidx.test.core.app.ApplicationProvider.getApplicationContext()),
                androidx.test.core.app.ApplicationProvider.getApplicationContext().getResources());
        assertEquals(expected, actual);
    }

    @Test
    public void testContentDescriptionTwoMinutes() {
        final String expected = "Expires in 2 minutes";
        final String actual = ExpirationTimer.getContentDescription(
                2,
                0,
                new TimeAgo(androidx.test.core.app.ApplicationProvider.getApplicationContext()),
                androidx.test.core.app.ApplicationProvider.getApplicationContext().getResources());
        assertEquals(expected, actual);
    }

    @Test
    public void testContentDescriptionThreeMinutesOneSecond() {
        final String expected = "Expires in 3 minutes and 1 second";
        final String actual = ExpirationTimer.getContentDescription(
                3,
                1,
                new TimeAgo(androidx.test.core.app.ApplicationProvider.getApplicationContext()),
                androidx.test.core.app.ApplicationProvider.getApplicationContext().getResources());
        assertEquals(expected, actual);
    }

    @Test
    public void testContentDescriptionSixMinutesFiftyNineSeconds() {
        final String expected = "Expires in 6 minutes and 59 seconds";
        final String actual = ExpirationTimer.getContentDescription(
                6,
                59,
                new TimeAgo(androidx.test.core.app.ApplicationProvider.getApplicationContext()),
                androidx.test.core.app.ApplicationProvider.getApplicationContext().getResources());
        assertEquals(expected, actual);
    }
}

File: app/src/test/java/com/launchkey/android/authenticator/sdk/ui/internal/view/ExpirationTimerTest.kt
/*
 *  Copyright (c) 2018. iovation, LLC. All rights reserved.
 */
package com.launchkey.android.authenticator.sdk.ui.internal.view

import android.content.Context
import androidx.test.core.app.ApplicationProvider
import com.launchkey.android.authenticator.sdk.ui.internal.common.TimeAgo
import com.launchkey.android.authenticator.sdk.ui.internal.view.ExpirationTimer.Companion.calculateMinSec
import com.launchkey.android.authenticator.sdk.ui.internal.view.ExpirationTimer.Companion.convertMinSecToFormattedString
import com.launchkey.android.authenticator.sdk.ui.internal.view.ExpirationTimer.Companion.getContentDescription
import org.junit.Assert
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

// Robolectric runner required to
// shadow a static call of Color.rgb() for
// default values in BaseExpirationTimer
@RunWith(RobolectricTestRunner::class)
@Config(sdk = [23])
class ExpirationTimerTest {
    @Test
    fun testMillisToMinSecNullCheck() {
        val minSec = calculateMinSec(100)
        Assert.assertNotNull(minSec)
    }
    
    @Test
    fun testMillisToMinSecLengthCheck() {
        val minSec = calculateMinSec(100)
        Assert.assertEquals(2, minSec.size.toLong())
    }
    
    @Test
    fun testMillisToMinSecNegativeMillis() {
        val minSec = calculateMinSec(-100)
        Assert.assertNotNull(minSec)
        Assert.assertEquals(0L, minSec[0])
        Assert.assertEquals(0L, minSec[1])
    }
    
    @Test
    fun testMillisToMinSecBarelyBelowSixtySeconds() {
        val minSec = calculateMinSec(59999)
        Assert.assertNotNull(minSec)
        Assert.assertEquals(0L, minSec[0])
        Assert.assertEquals(59L, minSec[1])
    }
    
    @Test
    fun testMillisToMinSecSixtySeconds() {
        val minSec = calculateMinSec(60000)
        Assert.assertNotNull(minSec)
        Assert.assertEquals(1L, minSec[0])
        Assert.assertEquals(0L, minSec[1])
    }
    
    @Test
    fun testMillisToMinSecBarelyAboveSixtySeconds() {
        val minSec = calculateMinSec(60001)
        Assert.assertNotNull(minSec)
        Assert.assertEquals(1L, minSec[0])
        Assert.assertEquals(0L, minSec[1])
    }
    
    @Test
    fun testConvertMinSecToFormattedStringRegular() {
        val timeFormatted = convertMinSecToFormattedString(ExpirationTimer.FORMAT_DEFAULT, 1, 59)
        Assert.assertNotNull(timeFormatted)
        Assert.assertEquals("01:59", timeFormatted)
    }
    
    @Test
    fun testConvertMinSecToFormattedStringFiftyNineSeconds() {
        val timeFormatted = convertMinSecToFormattedString(ExpirationTimer.FORMAT_DEFAULT, 0, 59)
        Assert.assertNotNull(timeFormatted)
        Assert.assertEquals("00:59", timeFormatted)
    }
    
    @Test
    fun testConvertMinSecToFormattedStringDisallowSixtySeconds() {
        val timeFormatted = convertMinSecToFormattedString(ExpirationTimer.FORMAT_DEFAULT, 0, 60)
        Assert.assertNotNull(timeFormatted)
        Assert.assertEquals("00:00", timeFormatted)
    }
    
    @Test
    fun testConvertMinSecToFormattedStringDisallowOverSixtySeconds() {
        val timeFormatted = convertMinSecToFormattedString(ExpirationTimer.FORMAT_DEFAULT, 0, 61)
        Assert.assertNotNull(timeFormatted)
        Assert.assertEquals("00:00", timeFormatted)
    }
    
    @Test
    fun testConvertMinSecToFormattedStringDisallowUnderZeroSeconds() {
        val timeFormatted = convertMinSecToFormattedString(ExpirationTimer.FORMAT_DEFAULT, 0, -1)
        Assert.assertNotNull(timeFormatted)
        Assert.assertEquals("00:00", timeFormatted)
    }
    
    @Test
    fun testContentDescriptionOneSecond() {
        val expected = "Expires in 1 second"
        val actual = getContentDescription(
            0,
            1,
            TimeAgo(ApplicationProvider.getApplicationContext()),
            ApplicationProvider.getApplicationContext<Context>().resources
        )
        Assert.assertEquals(expected, actual)
    }
    
    @Test
    fun testContentDescriptionFiveSeconds() {
        val expected = "Expires in 5 seconds"
        val actual = getContentDescription(
            0,
            5,
            TimeAgo(ApplicationProvider.getApplicationContext()),
            ApplicationProvider.getApplicationContext<Context>().resources
        )
        Assert.assertEquals(expected, actual)
    }
    
    @Test
    fun testContentDescriptionOneMinute() {
        val expected = "Expires in 1 minute"
        val actual = getContentDescription(
            1,
            0,
            TimeAgo(ApplicationProvider.getApplicationContext()),
            ApplicationProvider.getApplicationContext<Context>().resources
        )
        Assert.assertEquals(expected, actual)
    }
    
    @Test
    fun testContentDescriptionOneMinuteOneSecond() {
        val expected = "Expires in 1 minute and 1 second"
        val actual = getContentDescription(
            1,
            1,
            TimeAgo(ApplicationProvider.getApplicationContext()),
            ApplicationProvider.getApplicationContext<Context>().resources
        )
        Assert.assertEquals(expected, actual)
    }
    
    @Test
    fun testContentDescriptionOneMinuteFortySeconds() {
        val expected = "Expires in 1 minute and 40 seconds"
        val actual = getContentDescription(
            1,
            40,
            TimeAgo(ApplicationProvider.getApplicationContext()),
            ApplicationProvider.getApplicationContext<Context>().resources
        )
        Assert.assertEquals(expected, actual)
    }
    
    @Test
    fun testContentDescriptionTwoMinutes() {
        val expected = "Expires in 2 minutes"
        val actual = getContentDescription(
            2,
            0,
            TimeAgo(ApplicationProvider.getApplicationContext()),
            ApplicationProvider.getApplicationContext<Context>().resources
        )
        Assert.assertEquals(expected, actual)
    }
    
    @Test
    fun testContentDescriptionThreeMinutesOneSecond() {
        val expected = "Expires in 3 minutes and 1 second"
        val actual = getContentDescription(
            3,
            1,
            TimeAgo(ApplicationProvider.getApplicationContext()),
            ApplicationProvider.getApplicationContext<Context>().resources
        )
        Assert.assertEquals(expected, actual)
    }
    
    @Test
    fun testContentDescriptionSixMinutesFiftyNineSeconds() {
        val expected = "Expires in 6 minutes and 59 seconds"
        val actual = getContentDescription(
            6,
            59,
            TimeAgo(ApplicationProvider.getApplicationContext()),
            ApplicationProvider.getApplicationContext<Context>().resources
        )
        Assert.assertEquals(expected, actual)
    }
}